service/history/api/get_workflow_util.go

```diff
@@ -79,14 +78,25 @@ func GetOrPollMutableState(
 	if err != nil {
 		return nil, err
 	}
-	if request.CurrentBranchToken == nil {
-		request.CurrentBranchToken = response.CurrentBranchToken
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(response.GetVersionHistories())
+	if err != nil {
+		return nil, err
+	}
+	if request.GetVersionHistoryItem() == nil {
+		lastVersionHistoryItem, err := versionhistory.GetLastVersionHistoryItem(currentVersionHistory)
+		if err != nil {
+			return nil, err
+		}
+		request.VersionHistoryItem = lastVersionHistoryItem
 	}
-	if !bytes.Equal(request.CurrentBranchToken, response.CurrentBranchToken) {
+	// Use the latest event id + event version as the branch identifier. This pair is unique across clusters.
+	// We return the full version histories. Callers need to fetch the last version history item from current branch
+	// and use the last version history item in following calls.
```

Comment:
> Actually I am a bit confused, why GetMutableState need to have currentBranchToken or lastVersionHistoryItem in the request and perform this check? To me, GetMutableState is just a more advanced version of DescribeWorkflowExecution and more details fields are returned in the response. The concept of BranchChange should not exist here?

Because we have GetOrPollMutableState and this method only take in GetMutableStateRequest. So we use GetMutableStateRequest for poll mutable stable case. I prefer we separate them. I can create an issue for it.

> For PollMutableState, yeah, the concept make sense since user is following the mutable state update. But asking caller to figure out how the version history work seems a lot. Any reason preventing us from just returning the lastVersionHistoryItem or even just some opaque polling token here?

The version histories is already in the response. I can add another field for last version history item if you prefer. If we go with this, there will be duplicated data in the response.

Reply:
Prefer to separate the get and poll with re-design the APIs in a different PR.

Reply:
sounds good. Let's create an issue and refactor those api in a different PR. 

It seems like no one is using the versionhistory response from PollMutableState.

Reply:
https://github.com/temporalio/temporal/issues/4958